go/consensus/tendermint/abci/mux.go

```diff
@@ -398,6 +389,37 @@ func (mux *abciMux) InitChain(req types.RequestInitChain) types.ResponseInitChai
 
 	mux.logger.Debug("Genesis hook dispatch complete")
 
+	mux.RUnlock()
+
+	// TODO Matevz: It seems stateKeyGenesisRequest is still used in roothash's ForeignExecuteTx for generating new runtimes?
+	b, _ = req.Marshal()
+	mux.state.deliverTxTree.Set([]byte(stateKeyGenesisRequest), b)
+	mux.state.checkTxTree.Set([]byte(stateKeyGenesisRequest), b)
+
+	// Call InitChain() on all applications.
+	mux.logger.Debug("InitChain: initializing applications")
+	ctx := NewContext(ContextInitChain, mux.currentTime, mux.state)
+
+	for _, app := range mux.appsByLexOrder {
+		mux.logger.Debug("InitChain: calling InitChain on application",
+			"app", app.Name(),
+		)
+
+		if err = app.InitChain(ctx, req, st); err != nil {
+			mux.logger.Error("InitChain: fatal error in application",
+				"err", err,
+				"app", app.Name(),
+			)
+			panic("mux: InitChain: fatal error in application: '" + app.Name() + "': " + err.Error())
+		}
+	}
+
+	mux.logger.Debug("InitChain: initializing of applications complete", "num_collected_events", len(ctx.GetEvents()))
+
+	// Since returning emitted events doesn't work for InitChain() response yet,
+	// we store those and return them in BeginBlock().
+	mux.initChainEvents = ctx.GetEvents()
```

Comment:
This doesn't seem right, the events should be stored in persistent state (`mux.state.*TxTree`) otherwise a node crash will just erase events and cause divergence.